### ISSUE TITLE
Replace [[offset]] with [[span]]?

### ISSUE THREAD

**Issue by dj2** [gpuweb/gpuweb#1303]
Thinking about struct offsets. Currently we do something like:

```rust
[[block]] struct MaterialUniforms {
  [[offset(0)]] baseColorFactor : vec4<f32>;
  [[offset(16)]] metallicRoughnessFactor : vec2<f32>;
  [[offset(32)]] emissiveFactor : vec3<f32>;
  [[offset(44)]] occlusionStrength : f32;
};
```

What if we shortened that to something like:

```rust
[[block]] struct MaterialUniforms {
  baseColorFactor : vec4<f32>;
  [[span(16)]] metallicRoughnessFactor : vec2<f32>;
  [[span(16)]] emissiveFactor : vec3<f32>;
  occlusionStrength : f32;
};
```

Where, basically, the offsets start at zero with the first element. If the next element distance is the same as the size of the type (so 16 bytes for a `vec4<f32>` or 12 bytes for a `vec3<f32>`) you don't have to provide anything. In the case where the size is different (the `vec3`'s having `vec4` size above) you annotate with a `span(<size>)`. Basically, saying here is the real size of this element. We would still have to validate things are correct, but provides the clarity of offsets, without the verbosity I think.

Would cut down down the number of annotations (looking through Austins examples, none of the structs would be annotated) and it fixes up the issue with shuffling members around as the annotation is on the line it refers too and doesn't effect other lines. Also removes the ability to have random sized holes in your data structure.

Related to: #621 #561

**Comment by dj2** [comment]
We can either make it a validation error if the span doesn't fit what we expect for the given size (i.e. vec3 expanding to vec4) or we allow injected padding as long as it's a vec4 multiple (vec3 could be spanned to a (2*vec4))

**Comment by potree** [comment]
I prefer the span approach over the offset approach because it makes prototyping with frequent additions, removal and rearanging of members a bit more convenient. My favourite would be an approach without the need to specify exact offsets or sizes (but make it optional) and offer an API that lets users query the structs, members and offsets on host side. One of the reasons is that for me, offsets or spans don't matter in some of my use cases because the more complex structs will be fully initialized by compute shaders on GPU side. I'd just need to know the total size of a struct on host side to allocate sufficient memory.

**Comment by dj2** [comment]
Yea, with the sizes matching you could just do:

```rust
[[block]] struct Data {
  a : f32;
  b : f32;
  c : f32;
}
```

No need for span in that case as there is no wasted space.

**Comment by Kangz** [comment]
`[[span]]` could help, but I'm worried it is losing the most important type of compile-time assertion that's "I know this member is exactly at this offset in the struct" (which you can compare to your CPU code's pointer arithmetic or `ArrayBuffer` indexing).

**Comment by dj2** [comment]
You still have that assertion here, you just have to do the math? Either the type exactly fills the number of bytes, or there is a span required. If you really wanted you could do:

```rust
[[block]] struct Data {
  [[span(4)]] a : f32;
  [[span(4)]] b : f32;
  [[span(4)]] c : f32;
};
```
Then start counting at 0 and add the span each time you want to move to the next member. But, I think we should have it optional for when it completely covers the size. So it would really only come into play with `vec3`, possibly `vec2` and some matrix sizes.

**Comment by Kangz** [comment]
Doing the sum ourselves is a bit less convenient. I guess in the ideal world I'd like to have:

```
[[block, layout_for(uniform)]] struct S {
     a : f32;
     b: f32;
     [[offset(8)]] c : vec3<f32>;
};
```

make the compiler error out with a message like "offset for member c of struct S is at offset 16 and not offset 8 as requested".

**Comment by m-schuetz** [comment]
What about making device.createShaderModule return data about all bindings, structs, and their members including offsets and sizes/strides, so that developers can allocate and populate the buffers as needed?

**Comment by dj2** [comment]
I'm not sure what doing it that way buys you? You don't know how big the `vec3` is, you just know where it starts? It also suffers from the issue that if you add a `b2 : f32` after `b` but before `c` you have to adjust `c`.

**Comment by dneto0** [comment]
This idea looks like a good improvement:    It removes the "action at a distance" pain point, where changing an early member causes a cascade of offset updates.

**Comment by kvark** [comment]
The rough consensus today on the call was to have it, and the next step was @dj2 writing an actual PR.
Personally, I think this is great 👍

**Comment by litherum** [comment]
@dj2

> If the next element distance is the same as the size of the type (so 16 bytes for a vec4<f32> or 12 bytes for a vec3<f32>) you don't have to provide anything.

This sounds great.

@Kangz 

> make the compiler error out with a message like "offset for member c of struct S is at offset 16 and not offset 8 as requested".

This sounds great too.

@m-schuetz

> What about making device.createShaderModule return data

This also sounds great.

**Comment by kvark** [comment]
I believe, in light of the issues we discovered in this approach, it should be closed in favor of #1393 as a spiritual successor.

**Comment by kainino0x** [comment]
Superceded by #1393.
